File: src/githubclaw/agents/spawner.py
"""Agent CLI spawner.

Builds CLI commands from AgentDefinition + prompt file path and launches
agent subprocesses. Supports both Claude Code and Codex backends, with
overridable shell script templates.
"""

from __future__ import annotations

import logging
import os
import subprocess
from typing import TYPE_CHECKING

from githubclaw.constants import DEFAULT_AGENT_MAX_TURNS
from githubclaw.errors import AgentSpawnError

if TYPE_CHECKING:
    from pathlib import Path

    from githubclaw.agents.parser import AgentDefinition

logger = logging.getLogger(__name__)

# Default spawn command templates (used when no shell script override exists).
_DEFAULT_CLAUDE_TEMPLATE = (
    "claude -p"
    " --dangerously-skip-permissions"
    " --allowedTools {allowed_tools}"
    " --disallowedTools {disallowed_tools}"
    " --max-turns {max_turns}"
    ' --append-system-prompt-file "{prompt_file}"'
    ' "{task_prompt}"'
)

_DEFAULT_CODEX_TEMPLATE = (
    'cat "{prompt_file}" | codex exec - --approval-mode full-auto --sandbox off'
)


class AgentSpawner:
    """Spawns agent CLI processes from definitions and assembled prompts."""

    def __init__(
        self,
        repo_root: Path,
        max_turns: int = DEFAULT_AGENT_MAX_TURNS,
    ) -> None:
        self.repo_root = repo_root
        self.max_turns = max_turns

    def _get_spawn_script(self, backend: str) -> Path | None:
        """Return the overridable spawn script path if it exists."""
        if backend == "claude-code":
            script = self.repo_root / ".githubclaw" / "spawn_claude.sh"
        elif backend == "codex":
            script = self.repo_root / ".githubclaw" / "spawn_codex.sh"
        else:
            return None
        return script if script.exists() else None

    def _build_env(
        self,
        agent_def: AgentDefinition,
        prompt_file: Path,
        task_prompt: str,
        extra_env: dict[str, str] | None = None,
    ) -> dict[str, str]:
        """Build environment variables for the agent subprocess."""
        env = os.environ.copy()

        # Git author identity for this agent.
        env["GIT_AUTHOR_NAME"] = agent_def.git_author_name
        env["GIT_AUTHOR_EMAIL"] = agent_def.git_author_email
        env["GIT_COMMITTER_NAME"] = agent_def.git_author_name
        env["GIT_COMMITTER_EMAIL"] = agent_def.git_author_email

        # Tool permissions (for spawn script templates).
        tools = agent_def.active_tools
        env["ALLOWED_TOOLS"] = ",".join(tools.allowed) if tools.allowed else ""
        env["DISALLOWED_TOOLS"] = ",".join(tools.disallowed) if tools.disallowed else ""

        # Prompt and task info.
        env["PROMPT_FILE"] = str(prompt_file)
        env["TASK_PROMPT"] = task_prompt
        env["MAX_TURNS"] = str(self.max_turns)

        # Agent metadata.
        env["GITHUBCLAW_AGENT_TYPE"] = agent_def.name
        env["GITHUBCLAW_BACKEND"] = agent_def.backend
        env["GITHUBCLAW_REPO_ROOT"] = str(self.repo_root)

        if extra_env:
            env.update(extra_env)

        return env

    def _build_command_from_template(
        self,
        agent_def: AgentDefinition,
        prompt_file: Path,
        task_prompt: str,
    ) -> list[str]:
        """Build the CLI command using built-in templates (no shell script override)."""
        tools = agent_def.active_tools

        if agent_def.backend == "claude-code":
            allowed = ",".join(tools.allowed) if tools.allowed else ""
            disallowed = ",".join(tools.disallowed) if tools.disallowed else ""

            cmd_parts: list[str] = ["claude", "-p", "--dangerously-skip-permissions"]

            if allowed:
                cmd_parts.extend(["--allowedTools", allowed])
            if disallowed:
                cmd_parts.extend(["--disallowedTools", disallowed])

            cmd_parts.extend(["--max-turns", str(self.max_turns)])
            cmd_parts.extend(["--append-system-prompt-file", str(prompt_file)])
            cmd_parts.append(task_prompt)

            return cmd_parts

        elif agent_def.backend == "codex":
            # Codex uses piped stdin, so we need shell=True via the script path.
            # When there's no script override, we build a shell command string.
            return [
                "bash",
                "-c",
                f'cat "{prompt_file}" | codex exec - --approval-mode full-auto --sandbox off',
            ]

        else:
            raise AgentSpawnError(f"Unknown backend: {agent_def.backend!r}")

    def spawn(
        self,
        agent_def: AgentDefinition,
        prompt_file: Path,
        task_prompt: str,
        extra_env: dict[str, str] | None = None,
    ) -> subprocess.Popen[bytes]:
        """Spawn an agent subprocess.

        Args:
            agent_def: Parsed agent definition.
            prompt_file: Path to the assembled prompt temp file.
            task_prompt: The task prompt string (for Claude Code's positional arg).
            extra_env: Additional environment variables (e.g. platform API keys).

        Returns:
            subprocess.Popen handle for the running agent process.
        """
        env = self._build_env(agent_def, prompt_file, task_prompt, extra_env)

        # Check for user-provided spawn script override.
        spawn_script = self._get_spawn_script(agent_def.backend)

        if spawn_script is not None:
            # Use the overridable shell script. All config is passed via env vars.
            cmd: list[str] | str = ["bash", str(spawn_script)]
            logger.info(
                "Spawning %s agent (%s backend) via script: %s",
                agent_def.name,
                agent_def.backend,
                spawn_script,
            )
        else:
            # Use built-in command template.
            cmd = self._build_command_from_template(agent_def, prompt_file, task_prompt)
            logger.info(
                "Spawning %s agent (%s backend) via built-in template",
                agent_def.name,
                agent_def.backend,
            )

        logger.debug("Agent command: %s", cmd)
        logger.debug("Agent working directory: %s", self.repo_root)

        try:
            process = subprocess.Popen(
                cmd,
                cwd=self.repo_root,
                env=env,
                stdout=subprocess.PIPE,
                stderr=subprocess.PIPE,
            )
        except (OSError, FileNotFoundError) as exc:
            logger.error(
                "Failed to spawn %s agent (%s backend): %s",
                agent_def.name,
                agent_def.backend,
                exc,
            )
            raise AgentSpawnError(
                f"Could not start {agent_def.backend!r} CLI for agent "
                f"{agent_def.name!r}. Is the binary installed and on PATH? "
                f"Original error: {exc}"
            ) from exc

        logger.info(
            "Agent %s spawned with PID %d (backend=%s, timeout=%s)",
            agent_def.name,
            process.pid,
            agent_def.backend,
            agent_def.timeout or "global default",
        )

        return process

File: src/agents/spawner.rs
//! Agent CLI spawner.
//!
//! Builds the environment variables and command-line arguments needed to launch
//! an agent subprocess (e.g. `claude-code` or `codex`) with the correct
//! prompt, tool permissions, and git identity.

use std::collections::HashMap;
use std::path::{Path, PathBuf};

use crate::agents::parser::AgentDefinition;

/// Embedded gh wrapper script (compiled into the binary).
const GH_WRAPPER_SCRIPT: &str = include_str!("../../scripts/gh");

/// Spawns agent subprocesses with the correct environment and CLI flags.
pub struct AgentSpawner {
    repo_root: PathBuf,
    max_turns: u32,
}

impl AgentSpawner {
    /// Create a new spawner rooted at `repo_root` with the given turn limit.
    pub fn new(repo_root: impl AsRef<Path>, max_turns: u32) -> Self {
        Self {
            repo_root: repo_root.as_ref().to_path_buf(),
            max_turns,
        }
    }

    /// Build the full set of environment variables for an agent subprocess.
    ///
    /// Sets git identity, tool permissions, prompt/task info, GithubClaw
    /// metadata, root issue tracking, and gh wrapper PATH injection.
    /// Optionally merges caller-supplied `extra_env`.
    pub fn build_env(
        &self,
        agent_def: &AgentDefinition,
        prompt_file: &Path,
        task_prompt: &str,
        extra_env: Option<&HashMap<String, String>>,
    ) -> HashMap<String, String> {
        let mut env = HashMap::new();
        let prompt_text = std::fs::read_to_string(prompt_file).unwrap_or_default();

        // Git identity
        env.insert("GIT_AUTHOR_NAME".into(), agent_def.git_author_name.clone());
        env.insert(
            "GIT_AUTHOR_EMAIL".into(),
            agent_def.git_author_email.clone(),
        );
        env.insert(
            "GIT_COMMITTER_NAME".into(),
            agent_def.git_author_name.clone(),
        );
        env.insert(
            "GIT_COMMITTER_EMAIL".into(),
            agent_def.git_author_email.clone(),
        );

        // Tool permissions
        let tools = agent_def.active_tools();
        env.insert("ALLOWED_TOOLS".into(), tools.allowed.join(","));
        env.insert("DISALLOWED_TOOLS".into(), tools.disallowed.join(","));

        // Prompt and task
        env.insert(
            "PROMPT_FILE".into(),
            prompt_file.to_string_lossy().into_owned(),
        );
        env.insert("SYSTEM_PROMPT".into(), prompt_text.clone());
        env.insert("TASK_CONTEXT".into(), task_prompt.to_string());
        let task_prompt_value = if agent_def.backend == "codex" {
            prompt_text
        } else {
            task_prompt.to_string()
        };
        env.insert("TASK_PROMPT".into(), task_prompt_value);
        env.insert("MAX_TURNS".into(), self.max_turns.to_string());

        // GithubClaw metadata
        env.insert("GITHUBCLAW_AGENT_TYPE".into(), agent_def.name.clone());
        env.insert("GITHUBCLAW_BACKEND".into(), agent_def.backend.clone());
        env.insert(
            "GITHUBCLAW_REPO_ROOT".into(),
            self.repo_root.to_string_lossy().into_owned(),
        );

        // Root issue tracking for ref #N injection
        // GITHUBCLAW_ROOT_ISSUE is injected by the caller via extra_env

        // gh wrapper PATH injection
        // Prepend the scripts/ directory (containing gh renamed to gh)
        // to PATH so all gh CLI calls go through our wrapper.
        if let Some(wrapper_dir) = self.gh_wrapper_dir() {
            let current_path = std::env::var("PATH").unwrap_or_default();
            env.insert(
                "PATH".into(),
                format!("{}:{}", wrapper_dir.display(), current_path),
            );
            // Tell the wrapper where the real gh binary is
            if let Ok(real_gh) = which::which("gh") {
                env.insert(
                    "GITHUBCLAW_REAL_GH".into(),
                    real_gh.to_string_lossy().into_owned(),
                );
            }
        }

        // Merge extra env (caller overrides take precedence)
        if let Some(extra) = extra_env {
            for (k, v) in extra {
                env.insert(k.clone(), v.clone());
            }
        }

        env
    }

    /// Get or create the gh wrapper directory.
    ///
    /// Extracts the embedded gh wrapper script to a stable temp directory
    /// so it's always available regardless of where the binary is installed.
    /// The wrapper is placed at `~/.githubclaw/bin/gh`.
    fn gh_wrapper_dir(&self) -> Option<PathBuf> {
        let wrapper_dir = crate::config::global_config_dir().join("bin");
        let wrapper_path = wrapper_dir.join("gh");

        // Only write if missing or outdated
        if !wrapper_path.exists() {
            if std::fs::create_dir_all(&wrapper_dir).is_err() {
                return None;
            }
            if std::fs::write(&wrapper_path, GH_WRAPPER_SCRIPT).is_err() {
                return None;
            }
            #[cfg(unix)]
            {
                use std::os::unix::fs::PermissionsExt;
                let _ =
                    std::fs::set_permissions(&wrapper_path, std::fs::Permissions::from_mode(0o755));
            }
        }

        Some(wrapper_dir)
    }

    /// Build the command-line arguments for launching the agent.
    ///
    /// Returns a `Vec<String>` where the first element is the program and the
    /// rest are arguments.
    ///
    /// # Errors
    ///
    /// Returns an error if the backend is not recognized.
    pub fn build_command(
        &self,
        agent_def: &AgentDefinition,
        _prompt_file: &Path,
        _task_prompt: &str,
    ) -> Result<Vec<String>, String> {
        self.build_inline_command(&agent_def.backend, false)
    }

    /// Build the command for a resumed Claude orchestrator session.
    pub fn build_resume_command(&self, agent_def: &AgentDefinition) -> Result<Vec<String>, String> {
        self.build_inline_command(&agent_def.backend, true)
    }

    fn build_inline_command(
        &self,
        backend: &str,
        include_resume: bool,
    ) -> Result<Vec<String>, String> {
        let shell = match backend {
            "claude-code" => {
                let mut command = concat!(
                    "cat \"$PROMPT_FILE\" | ",
                    "claude -p ",
                    "--dangerously-skip-permissions ",
                    "--allowedTools \"$ALLOWED_TOOLS\" ",
                    "--disallowedTools \"$DISALLOWED_TOOLS\" ",
                    "--max-turns \"$MAX_TURNS\""
                )
                .to_string();
                if include_resume {
                    command.push_str(" --resume \"$GITHUBCLAW_SESSION_NAME\"");
                }
                command
            }
            "codex" => concat!(
                "cat \"$PROMPT_FILE\" | ",
                "codex exec - ",
                "--dangerously-bypass-approvals-and-sandbox"
            )
            .to_string(),
            other => return Err(format!("unknown backend: {}", other)),
        };

        Ok(vec!["bash".to_string(), "-lc".to_string(), shell])
    }

    // spawn() will be async in the real implementation.
    // For now we only expose build_env and build_command for testing.
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::agents::parser::{AgentDefinition, ToolPermissions};
    use crate::constants::DEFAULT_AGENT_MAX_TURNS;
    use std::collections::HashMap;
    use tempfile::TempDir;

    /// Helper: build a minimal agent definition with sensible defaults.
    fn make_agent_def(backend: &str) -> AgentDefinition {
        let mut tools = HashMap::new();
        tools.insert(
            backend.to_string(),
            ToolPermissions {
                allowed: vec!["Read".into(), "Write".into()],
                disallowed: vec!["Shell".into()],
            },
        );
        AgentDefinition {
            name: "coder".into(),
            backend: backend.into(),
            git_author_name: "Test Bot".into(),
            git_author_email: "bot@test.local".into(),
            timeout: None,
            tools,
            instruction_body: "Do the work.".into(),
        }
    }

    // 1. build_env sets git author fields
    #[test]
    fn build_env_sets_git_author_fields() {
        let tmp = TempDir::new().unwrap();
        let spawner = AgentSpawner::new(tmp.path(), DEFAULT_AGENT_MAX_TURNS);
        let def = make_agent_def("claude-code");
        let prompt = tmp.path().join("prompt.md");

        let env = spawner.build_env(&def, &prompt, "fix bug", None);

        assert_eq!(env["GIT_AUTHOR_NAME"], "Test Bot");
        assert_eq!(env["GIT_AUTHOR_EMAIL"], "bot@test.local");
        assert_eq!(env["GIT_COMMITTER_NAME"], "Test Bot");
        assert_eq!(env["GIT_COMMITTER_EMAIL"], "bot@test.local");
    }

    // 2. build_env sets tool permissions
    #[test]
    fn build_env_sets_tool_permissions() {
        let tmp = TempDir::new().unwrap();
        let spawner = AgentSpawner::new(tmp.path(), 100);
        let def = make_agent_def("claude-code");
        let prompt = tmp.path().join("prompt.md");
        std::fs::write(&prompt, "System prompt text").unwrap();

        let env = spawner.build_env(&def, &prompt, "task", None);

        assert_eq!(env["ALLOWED_TOOLS"], "Read,Write");
        assert_eq!(env["DISALLOWED_TOOLS"], "Shell");
        assert_eq!(env["SYSTEM_PROMPT"], "System prompt text");
        assert_eq!(env["TASK_CONTEXT"], "task");
    }

    // 3. build_env sets prompt and task info
    #[test]
    fn build_env_sets_prompt_and_task_info() {
        let tmp = TempDir::new().unwrap();
        let spawner = AgentSpawner::new(tmp.path(), 50);
        let def = make_agent_def("codex");
        let prompt = tmp.path().join("prompt.md");
        std::fs::write(&prompt, "full prompt").unwrap();

        let env = spawner.build_env(&def, &prompt, "implement caching", None);

        assert_eq!(env["PROMPT_FILE"], prompt.to_string_lossy().as_ref());
        assert_eq!(env["SYSTEM_PROMPT"], "full prompt");
        assert_eq!(env["TASK_CONTEXT"], "implement caching");
        assert_eq!(env["TASK_PROMPT"], "full prompt");
        assert_eq!(env["MAX_TURNS"], "50");
        assert_eq!(env["GITHUBCLAW_AGENT_TYPE"], "coder");
        assert_eq!(env["GITHUBCLAW_BACKEND"], "codex");
        assert_eq!(
            env["GITHUBCLAW_REPO_ROOT"],
            tmp.path().to_string_lossy().as_ref()
        );
    }

    // 4. build_env merges extra_env
    #[test]
    fn build_env_merges_extra_env() {
        let tmp = TempDir::new().unwrap();
        let spawner = AgentSpawner::new(tmp.path(), 100);
        let def = make_agent_def("claude-code");
        let prompt = tmp.path().join("prompt.md");
        std::fs::write(&prompt, "System prompt text").unwrap();

        let mut extra = HashMap::new();
        extra.insert("CUSTOM_VAR".into(), "custom_value".into());
        extra.insert("GIT_AUTHOR_NAME".into(), "Override Bot".into());

        let env = spawner.build_env(&def, &prompt, "task", Some(&extra));

        assert_eq!(env["CUSTOM_VAR"], "custom_value");
        // Extra env overrides built-in values
        assert_eq!(env["GIT_AUTHOR_NAME"], "Override Bot");
    }

    // 5. build_command for claude-code backend
    #[test]
    fn build_command_claude_code() {
        let tmp = TempDir::new().unwrap();
        let spawner = AgentSpawner::new(tmp.path(), 200);
        let def = make_agent_def("claude-code");
        let prompt = tmp.path().join("prompt.md");

        let cmd = spawner.build_command(&def, &prompt, "fix the bug").unwrap();

        assert_eq!(cmd[0], "bash");
        assert_eq!(cmd[1], "-lc");
        assert!(cmd[2].contains("cat \"$PROMPT_FILE\" | claude -p"));
        assert!(cmd[2].contains("--max-turns \"$MAX_TURNS\""));
        assert!(!cmd[2].contains("--prompt-file"));
        assert!(!cmd[2].contains("--task"));
    }

    // 6. build_command for codex backend
    #[test]
    fn build_command_codex() {
        let tmp = TempDir::new().unwrap();
        let spawner = AgentSpawner::new(tmp.path(), 100);
        let def = make_agent_def("codex");
        let prompt = tmp.path().join("prompt.md");

        let cmd = spawner
            .build_command(&def, &prompt, "implement feature")
            .unwrap();

        assert_eq!(cmd[0], "bash");
        assert_eq!(cmd[1], "-lc");
        assert!(cmd[2].contains("cat \"$PROMPT_FILE\" | codex exec -"));
        assert!(cmd[2].contains("--dangerously-bypass-approvals-and-sandbox"));
        assert!(!cmd[2].contains("--prompt-file"));
        assert!(!cmd[2].contains("--task"));
    }

    // 7. build_command for unknown backend returns error
    #[test]
    fn build_command_unknown_backend_errors() {
        let tmp = TempDir::new().unwrap();
        let spawner = AgentSpawner::new(tmp.path(), 100);
        let def = make_agent_def("unknown-backend");
        let prompt = tmp.path().join("prompt.md");

        let result = spawner.build_command(&def, &prompt, "task");
        assert!(result.is_err());
        assert!(result.unwrap_err().contains("unknown backend"));
    }

    #[test]
    fn build_command_ignores_repo_local_spawn_scripts() {
        let tmp = TempDir::new().unwrap();
        let claw_dir = tmp.path().join(".githubclaw");
        std::fs::create_dir_all(&claw_dir).unwrap();

        let script_path = claw_dir.join("spawn_claude.sh");
        std::fs::write(&script_path, "#!/bin/bash\nexit 99").unwrap();

        let spawner = AgentSpawner::new(tmp.path(), 200);
        let def = make_agent_def("claude-code");
        let prompt = tmp.path().join("prompt.md");

        let cmd = spawner.build_command(&def, &prompt, "fix the bug").unwrap();
        assert_eq!(cmd[0], "bash");
        assert_eq!(cmd[1], "-lc");
        assert!(cmd[2].contains("cat \"$PROMPT_FILE\" | claude -p"));
    }

    #[test]
    fn build_resume_command_claude_adds_resume_flag() {
        let tmp = TempDir::new().unwrap();
        let spawner = AgentSpawner::new(tmp.path(), 42);
        let def = make_agent_def("claude-code");

        let cmd = spawner.build_resume_command(&def).unwrap();

        assert_eq!(cmd[0], "bash");
        assert_eq!(cmd[1], "-lc");
        assert!(cmd[2].contains("--resume \"$GITHUBCLAW_SESSION_NAME\""));
        assert!(!cmd[2].contains("--prompt-file"));
    }
}
